src/runtime/formal/ops/reduce.cc

```diff
@@ -110,50 +79,28 @@ static void Reduce(CVMArgs args, reduce_func const& f) {
      *  2. foreach reduce axis dimension(n2, n3), 
      *      get x possible indexes and add value to result.
      **/
-//<<<<<<< HEAD
-//    for(uint64_t i = 0; i < getSize(y); i++){
-//      // for each y to be filled, find related xs and calculate.
-//      // the index for each dim of an x:
-//      // for a dim to be reduced, index of this dim differs from each x.
-//      // otherwise, it is fixed with y during the traverse.
-//      TShape yIndex = VectorIndex(reducedYShape, i);
-//      TShape xIndex(x->ndim);
-//      for (uint j = 0, yi = 0; j < x->ndim; j++) {
-//        xIndex[j] = flag[j] ? 0 : yIndex[yi++];
-//      }
-//      // the first x is tmp.
-//      int32_t tmp = x_data[ScalarIndex(xShape, xIndex)];
-//      for (uint64_t xi = 1; xi < axis_size; xi++) {
-//        TShape reducedIndex = VectorIndex(reducedAxShape, xi);
-//        for (uint j = 0, yi = 0, ri = 0; j < xIndex.ndim(); j++) {
-//          xIndex[j] = flag[j] ? reducedIndex[ri++] : yIndex[yi++];
-//        }
-//        f(tmp, x_data[ScalarIndex(xShape, xIndex)]);
-//=======
-    for(uint64_t i = 0; i < Y_shape.Size(); i++){
-      // in_i will be the base index of X. for Y[a][b] = sum(or max) X[a][*][*][d]
-      // in_i = d * 1 + a * n4*n3*n2
-      // o_i is a middle variable used for calculating in_i
-      uint64_t in_i = 0, o_i = i;
-      for(int j = yndim-1, xj = X_shape.ndim()-1; j>=0; j--){
-        uint64_t col = o_i % yshape[j];
-        o_i /= yshape[j];
-        while(xj >= 0 && flag[xj--]); // xj+1 is the dim that remains
-        in_i += col * every_xdim_size[xj+1];
-      }
 
-      int32_t tmp = X_data[in_i];
-      for(uint64_t xi = 1; xi < axis_size; xi++){
-        uint64_t o_i = xi, tmp_in_i = 0;
-        for(int j = realAxis.size() - 1; j>=0; j--){
-          uint64_t col = o_i % X_shape[realAxis[j]];
-          o_i /= X_shape[realAxis[j]];
-          tmp_in_i += col * every_xdim_size[realAxis[j]];
+    for (Indices yIdx(reducYShape); !yIdx.End(); yIdx++) {
+      // for each y to be filled, find related xs and calculate.
+      // the index for each dim of an x:
+      // for a dim to be reduced, index of this dim differs from each x.
+      // otherwise, it is fixed with y during the traverse.
+      Indices xIdx(X_shape);
```

Comment:
Move the `Indices` instantiation out of the loop or construct the `xIdx` in the for loop's first condition:

```suggestion
    for (Indices yIdx(reducYShape), xIdx(X_shape); !yIdx.End(); yIdx++) {
      // for each y to be filled, find related xs and calculate.
      // the index for each dim of an x:
      // for a dim to be reduced, index of this dim differs from each x.
      // otherwise, it is fixed with y during the traverse.
      xIdx.reset();
```